File: algorithms/912-sort-an-array/sort-an-array.py
"""
912. sort an array

https://leetcode.com/problems/sort-an-array/
https://leetcode-cn.com/problems/sort-an-array/

排序数组
"""

from ast import Num
from operator import le
from tkinter import N
from typing import List
import random


class Solution:
    """
    快速排序 quick sort
    """

    def sort_array(self, nums: List[int]) -> List[int]:
        self.randomized_quicksort(nums, 0, len(nums) - 1)
        return nums

    def randomized_partition(self, nums: List[int], l: int, r: int):
        pivot = random.randint(l, r)
        nums[pivot], nums[r] = nums[r], nums[pivot]
        i = l - 1
        for j in range(l, r):
            if nums[j] < nums[r]:
                i += 1
                nums[i], nums[j] = nums[j], nums[i]
        i += 1
        nums[i], nums[r] = nums[r], nums[i]
        return i

    def randomized_quicksort(self, nums, l, r):
        if r <= l:
            return
        mid = self.randomized_partition(nums, l, r)
        self.randomized_quicksort(nums, l, mid - 1)
        self.randomized_quicksort(nums, mid + 1, r)


class SolutionV2:
    """
    归并排序 merge sort
    """

    def sort_array(self, nums: List[int]) -> List[int]:
        self.merge_sort(nums, 0, len(nums) - 1)
        return nums

    def merge_sort(self, nums: List[int], left: int, right: int):
        if left == right:
            return
        mid = left + (right - left) // 2
        self.merge_sort(nums, left, mid)
        self.merge_sort(nums, mid + 1, right)
        i, j = left, mid + 1
        tmp = []
        while i <= mid or j <= right:
            if i > mid or (j <= right and nums[j] < nums[i]):
                tmp.append(nums[j])
                j += 1
            else:
                tmp.append(nums[i])
                i += 1
        nums[left:right + 1] = tmp

File: algorithms/912-sort-an-array/sort-an-array.go
package problem912

import (
	"math/rand"
)

// 快速排序
func sortArray(nums []int) []int {
	randomizedQuickSort(nums, 0, len(nums)-1)
	return nums
}

func randomizedQuickSort(nums []int, left, right int) {
	if right <= left {
		return
	}
	mid := randomizedPartition(nums, left, right)
	randomizedQuickSort(nums, left, mid-1)
	randomizedQuickSort(nums, mid+1, right)
}

func randomizedPartition(nums []int, left, right int) int {
	pivot := rand.Intn(right-left) + left // 区间中取随机数
	nums[pivot], nums[right] = nums[right], nums[pivot]
	i := left - 1
	for j := left; j < right; j++ {
		if nums[j] < nums[right] {
			i++
			nums[i], nums[j] = nums[j], nums[i]
		}
	}
	i++
	nums[i], nums[right] = nums[right], nums[i]
	return i
}

// 归并排序
func sortArrayV2(nums []int) []int {
	mergeSort(nums, 0, len(nums)-1)
	return nums
}

// 合并
func mergeSort(nums []int, left, right int) {
	if right <= left {
		return
	}
	mid := (left + right) / 2
	mergeSort(nums, left, mid)
	mergeSort(nums, mid+1, right)
	var tmp []int
	i, j := left, mid+1
	for i <= mid || j <= right {
		if i > mid || (j <= right && nums[j] < nums[i]) {
			tmp = append(tmp, nums[j])
			j++
		} else {
			tmp = append(tmp, nums[i])
			i++
		}
	}

	copy(nums[left:right+1], tmp)
}
